### PR TITLE
Framework for running Probes in ClusterLoader2

### PR DESCRIPTION
**Pull request by mm4tt**
Probes definition can be found in https://github.com/kubernetes/perf-tests/pull/474

The measurement is currently disabled, will be enabling it gradually.

Ref. https://github.com/kubernetes/kubernetes/issues/76012

### CODE REVIEW

File: clusterloader2/pkg/measurement/common/probes/manifests/ping-server-serviceMonitor.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/manifests/ping-server-serviceMonitor.yaml]: Why prometheus need to monitor ping-server?

File: clusterloader2/pkg/measurement/common/probes/manifests/ping-server-service.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: ping-server
+  labels:
+    probe: ping-server
+spec:
+
+  ports:
+    - name: metrics
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/manifests/ping-server-service.yaml]: Does server expose metrics? I thought it isn't

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+	checkProbesReadyInterval = 30 * time.Second
+	checkProbesReadyTimeout  = 5 * time.Minute
+
+	currentProbesMetricsVersion = "v1"
+)
+
+var probesEnabled bool
+
+func init() {
+	flags.BoolEnvVar(&probesEnabled, "enable-probes", "ENABLE_PROBES", false, "Whether to run the probes in the clusterloader2 tests.")
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: Why is this needed?
You need to explicitly perform actions on this measurement, so I don't think this additional guard really provides value.

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+	// createProbesMeasurement method is the only place in go code that needs to be changed while
+	// adding a new probe.
+	// Each query template should accept a single %v placeholder corresponding to the query window
+	// length. See the 'in_cluster_network_latency' as an example.
+	probeNameToPrometheusQueryTmpl map[string]string
+
+	framework        *framework.Framework
+	replicasPerProbe int
+	templateMapping  map[string]interface{}
+	nServiceMonitors int
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: What is nServiceMonitors?

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+	startTime        time.Time
+	mutex            sync.Mutex
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: What is mutex protecting?

i think that in all other measurements there was silent assumption that the Execute method is not thread-safe and we don't do any locking. So if this was the motivation here, I would suggest removing this for now (just for consistency).

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+	}
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "start":
+		klog.Info("Starting probes...")
+		err := p.start(config)
+		return nil, err
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: return nil, p.start(config)

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+		klog.Info("Probes are disabled, skipping the measurement execution")
+		return nil, nil
+	}
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "start":
+		klog.Info("Starting probes...")
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: Can you move this to start() method?

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+		err := p.start(config)
+		return nil, err
+	case "gather":
+		klog.Info("Gathering metrics from probes...")
+		summary, err := p.gather(config.Params)
+		if err != nil && !errors.IsMetricViolationError(err) {
+			return nil, err
+		}
+		return []measurement.Summary{summary}, nil
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: s/nil/err/ ?

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+	}
+	klog.Info("Stopping probes...")
+	k8sClient := p.framework.GetClientSets().GetClient()
+	if err := client.DeleteNamespace(k8sClient, probesNamespace); err != nil {
+		klog.Errorf("error while deleting %s namespace: %v", probesNamespace, err)
+	}
+	if err := client.WaitForDeleteNamespace(k8sClient, probesNamespace); err != nil {
+		klog.Errorf("error while waiting for %s namespace to be deleted: %v", probesNamespace, err)
+	}
+	*p = *createProbesMeasurement().(*probesMeasurement) // Reset measurement state.
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: I don't think that's needed - we don't need to reinitialize it becasue the same measurement instantiation won't be reused.

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+				err = errors.NewMetricViolationError(probe.name, err.Error())
+				klog.Errorf("%s: %v", p, err)
+			}
+		}
+		probeSummaries = append(probeSummaries, probe)
+	}
+	return &probesSummary{probeSummaries: probeSummaries}, err
+}
+
+func (p *probesMeasurement) createProbesObjects() error {
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: Seems really similar to this one:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/prometheus/prometheus.go#L134

Can we refactor and reuse?

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+				"Perc99": float64(probe.latency.Perc99) / 1000000,
+			},
+			Unit:   "ms",
+			Labels: map[string]string{"Metric": name},
+		})
+	}
+	return util.PrettyPrintJSON(perfData)
+}
+
+// TODO(mm4tt): Remove the method below and start using the one from common util once it's available.
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: What is this blocked on? Can we do that now?

File: clusterloader2/testing/density/config.yaml [hunk 12]
```diff
@@ -54,6 +54,11 @@ steps:
       action: start
       nodeMode: {{$NODE_MODE}}
       resourceConstraints: {{$DENSITY_RESOURCE_CONSTRAINTS_FILE}}
+  - Identifier: Probes
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: Can we enable in a separate PR? In case of problems it's smaller thing to revert then :-)

File: clusterloader2/pkg/measurement/common/probes/manifests/ping-server-serviceMonitor.yaml (hunk 1 above)
Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/probes/manifests/ping-server-serviceMonitor.yaml]: OK - let's follow up on this on the other PR.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 3 above)
Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: OK - I got your point. Kubemark is definitely a valid example.

I think it's a different thing that chaos-monkey - it's a measurement that is needed for SLO measuring. Let me think about it a bit more.

File: clusterloader2/testing/density/config.yaml (hunk 12 above)
Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: Still - even if unused I would prefer to add it in a separate PR, you still register it, even if it doesn't perform any non-trivial code, it does something.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 10 above)
Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Done in https://github.com/kubernetes/perf-tests/pull/482

File: clusterloader2/testing/density/config.yaml (hunk 12 above)
Comment (mm4tt) [reply clusterloader2/testing/density/config.yaml]: Done

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 3 above)
Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: I was thinking a bit about it and my current thinking about it is:

Kubemark is quite specific setup - some things just doesn't make sense. I tend to think that at some point we may be forced to fork our tests for kubemark, but obviously I would like to delay it by as much as possible.

This is the same case for measurements - some just may not make sense in Kubemark. And this is exactly one example.
And in my opinion such things should be handle directly in measurement - if measurement doesn't make sense in Kubemark, it should just print some warning or sth like that and not start. This will solve the problem of kubemark.

Going to the problem of disabling (or gradually enabling) stuff - I don't think it's the right approach. We should try to keep all the tests consistent. There is a question how we should enable stuff, but how I see that is that we may create a temporary fork of config and have an "experimental" job that will be using that config. And once we believe it works, we will be graduating it to the main config (for all setups).

Thoughts?

Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Sounds good. Added a code that disables probes in kubemark and will enable the probes gradually via templates. 
Thanks!

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+		return err
+	}
+	k8sClient := p.framework.GetClientSets().GetClient()
+	if err := client.CreateNamespace(k8sClient, probesNamespace); err != nil {
+		return err
+	}
+	if err := p.createProbesObjects(); err != nil {
+		return err
+	}
+	if err := p.waitTillProbesReady(); err != nil {
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: waitForProbesReady ?

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+			if err != nil {
+				return nil, err
+			}
+			latency := time.Duration(float64(sample.Value) * float64(time.Second))
+			probe.latency.SetQuantile(quantile, latency)
+		}
+		if threshold, ok := thresholds[probeName]; ok {
+			if err := probe.latency.VerifyThreshold(threshold); err != nil {
+				err = errors.NewMetricViolationError(probe.name, err.Error())
+				klog.Errorf("%s: %v", p, err)
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: I think we don't want this error - this will be logged when checking violations above, right?

File: clusterloader2/pkg/measurement/common/probes/probes.go [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+}
+
+type probeSummary struct {
+	name    string
+	latency measurementutil.LatencyMetric
+}
+
+// SummaryName returns name of the summary.
+func (p *probesSummary) SummaryName() string {
+	return name
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/probes/probes.go]: This isn't very intuitive - I think that different probes are conceptually different things.
As a user, I would more expect to be separate summaries.

Why not simply have just probeSummary and return a separate one for every single probe?

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 11 above)
Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: What is the status of it?

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 13 above)
Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Done.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 15 above)
Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: The motivation here was to group all the probes together and export in a single, unified form, so they can be easily (and generically) imported into PerfDash. My aim was to make adding a new probe as seamless as possible (in this case no changes required in the PerfDash at all). 

But I can change it and export each probe separately if you think that's better. PLMK.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 14 above)
Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Good catch, now should be good.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 11 above)
Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: @krzysied, any updates?

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 15 above)
Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: I think the goal is great as mentioned, and I think you're already roughly doing what I want:
- probe name is already in probeNameToPrometheusQueryTmpl and is propagate down to probeSummary
- you already have this "probeSummary" struct

I'm just suggesting to:
- remove `probesSummary` and leave only `probeSummary`
- the methods on probesSummary will now become methos on probeSummary
- return []probeSummary as the result (separate summary for every probe)

Does that make sense?

Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Done.

Comment (wojtek-t) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Partially - I thought it should be p.Name here.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 11 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: I haven't touch it yet.
Assuming that for some reason it is blocking I will look into this today.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 15 above)
Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Doh! Thanks, fixed.

**Comment by mm4tt** [comment]
/hold

A lot has changed since I've tested it locally last time. Putting a hold on this PR to remember to retest before merging.

File: clusterloader2/pkg/measurement/common/probes/probes.go (hunk 11 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: https://github.com/kubernetes/perf-tests/pull/502

Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Thanks, will wait till https://github.com/kubernetes/perf-tests/pull/502 gets merged and remove this code.

Comment (mm4tt) [reply clusterloader2/pkg/measurement/common/probes/probes.go]: Done.

**Comment by wojtek-t** [comment]
/lgtm
/approve

Please cancel hold once you will be ready.

**Comment by mm4tt** [comment]
> New changes are detected. LGTM label has been removed.

I haven't changed anything... :)

**Comment by mm4tt** [comment]
/hold cancel
Run some tests, everything seems to be working fine.

@wojtek-t, I haven't changed anything but Prow Robot removed your lgtm, could you lgtm again? :)
